src/System.Management.Automation/engine/MshObject.cs

```diff
@@ -2347,6 +2378,204 @@ internal bool IsHelpObject
         #endregion
     }
 
+    /// <summary>
+    /// A factory class for more efficient creation of PSCustomObjects
+    /// </summary>
+    /// <para/>
+    /// This creates a context in which the guarantee is given that the
+    /// object being created, and the members that are being added are not
+    /// visible from any other location.
+    /// This way, defensive copies and otherwise expensive operations can be avoided.
+    public ref struct PSObjectBuilder
+    {
+        const int DefaultHirearchyDepth = 4;
+        private readonly HashSet<string> _validatedNames;
+        private Stack<PSObject> _currentObjectStack;
+
+        private PSObject _currentObject;
+
+        private Stack<PSObject> CurrentObjectStack => _currentObjectStack ?? (_currentObjectStack = new Stack<PSObject>(DefaultHirearchyDepth));
+
+        /// <summary>
+        /// Initializes a new instance of PSObjectBuilder.
+        /// </summary>
+        /// <param name="validatedNameCount">The initial size of the set that tracks validated parameter names.</param>
+        public PSObjectBuilder(int validatedNameCount)
+        {
+            if (validatedNameCount < 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(validatedNameCount));
+
+            }
+            _validatedNames = new HashSet<string>(validatedNameCount);
+            _currentObjectStack = null;
+            _currentObject = null;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of PSObjectBuilder.
+        /// </summary>
+        /// <param name="validatedNameCount">The initial size of the set that tracks validated parameter names.</param>
+        /// <param name="trustedNames">True if all the names are preValidated.</param>
+        internal PSObjectBuilder(int validatedNameCount, bool trustedNames)
+        {
+            if (validatedNameCount < 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(validatedNameCount));
+
+            }
+
+            _validatedNames = null;
+            _currentObjectStack = null;
+            _currentObject = null;
+        }
+
+        /// <summary>
+        /// Begins the creation of a new PSObject
+        /// </summary>
+        /// <param name="capacity">The initial capacity for the members collection.</param>
+        public void BeginCreateObject(int capacity)
+        {
+            if (_currentObject != null)
+            {
+                CurrentObjectStack.Push(_currentObject);
+            }
+            _currentObject = new PSObject(capacity);
```

Comment:
Can we by default push the current object on stack? Seem in the case we could initilize  _currentObject in constructor and exclude many  `if (_currentObject != null)`.